platform-dependent-components/problem-solver/cxx/message-classification-module/classifier/MessageTopicClassifier.cpp

```diff
@@ -0,0 +1,381 @@
+
+namespace messageClassificationModule
+{
+MessageTopicClassifier::MessageTopicClassifier(ScMemoryContext * context, WitAiClientInterface * client)
+  : context(context)
+  , client(client)
+{
+  messageSearcher = std::make_unique<MessageSearcher>(this->context);
+}
+
+ScAddrVector MessageTopicClassifier::classifyMessage(ScAddr const & messageAddr)
+{
+  ScAddrVector messageClassificationElements = {messageAddr};
+
+  std::string const messageText = getMessageText(messageAddr);
+
+  json const witResponse = client->getWitResponse(messageText);
+  SC_LOG_INFO(witResponse);
+
+  ScAddrVector const messageIntentElements = getMessageIntentClass(messageAddr, witResponse);
+  messageClassificationElements.insert(
+      messageClassificationElements.cend(), messageIntentElements.cbegin(), messageIntentElements.cend());
+
+  ScAddrVector const messageTraitElements = getMessageTraitClass(messageAddr, witResponse);
+  messageClassificationElements.insert(
+      messageClassificationElements.cend(), messageTraitElements.cbegin(), messageTraitElements.cend());
+
+  ScAddrVector const messageEntitiesElements = getMessageEntity(messageAddr, witResponse);
+  messageClassificationElements.insert(
+      messageClassificationElements.cend(), messageEntitiesElements.cbegin(), messageEntitiesElements.cend());
+
+  return messageClassificationElements;
+}
+
+std::string MessageTopicClassifier::getMessageText(ScAddr const & messageAddr)
+{
+  ScAddr const messageLink = messageSearcher->getMessageLink(messageAddr);
+  if (!messageLink.IsValid())
+  {
+    SC_THROW_EXCEPTION(utils::ExceptionItemNotFound, "MessageTopicClassifier: Message link is not found.");
+  }
+  std::string linkContent;
+  if (context->GetLinkContent(messageLink, linkContent))
+  {
+    return linkContent;
+  }
+  return {};
+}
+
+ScAddrVector MessageTopicClassifier::getMessageIntentClass(ScAddr const & messageAddr, json const & witResponse)
+{
+  ScAddrVector messageIntentCLassElements;
+
+  std::string const messageIntent = getMessageIntent(witResponse);
+  if (messageIntent.empty())
+  {
+    ScAddr const & messageIntentCLassEdge = context->CreateEdge(
+        ScType::EdgeAccessConstPosPerm,
+        MessageClassificationKeynodes::concept_not_classified_by_intent_message,
+        messageAddr);
+    messageIntentCLassElements.push_back(MessageClassificationKeynodes::concept_not_classified_by_intent_message);
+    messageIntentCLassElements.push_back(messageIntentCLassEdge);
+    return messageIntentCLassElements;
+  }
+
+  ScIterator3Ptr const possibleIntentIterator = context->Iterator3(
+      MessageClassificationKeynodes::concept_intent_possible_class,
+      ScType::EdgeAccessConstPosPerm,
+      ScType::NodeConstClass);
+
+  std::vector<std::string> witAiIdtfs;
+  ScAddr possibleMessageCLass;
+  while (possibleIntentIterator->Next())
+  {
+    possibleMessageCLass = possibleIntentIterator->Get(2);
+    witAiIdtfs = getWitAiIdtfs(possibleMessageCLass);
```

Comment:
possibly it will be more optimal to searh links with _messageIntent_ content, and the check that found link is a wit idtf of some class